File: go-cgo-impl/compute.go
package main

import (
	"sort"
	"sync"
	"time"
)

// ─── Quickselect helpers ────────────────────────────────────────────────────

// medianOfThree returns the index of the median of e[lo], e[mid], e[hi] by score.
func medianOfThree(e []scoreEntry, lo, mid, hi int) int {
	a, b, c := e[lo].score, e[mid].score, e[hi].score
	if (a >= b && b >= c) || (c >= b && b >= a) {
		return mid
	}
	if (b >= a && a >= c) || (c >= a && a >= b) {
		return lo
	}
	return hi
}

// partitionDesc partitions e[lo..hi] (inclusive) around a median-of-three pivot
// so that all entries left of the returned index have score >= pivot.score.
func partitionDesc(e []scoreEntry, lo, hi int) int {
	mid := lo + (hi-lo)/2
	pivotIdx := medianOfThree(e, lo, mid, hi)
	e[pivotIdx], e[hi] = e[hi], e[pivotIdx]
	pivot := e[hi].score

	i := lo - 1
	for j := lo; j < hi; j++ {
		if e[j].score >= pivot {
			i++
			e[i], e[j] = e[j], e[i]
		}
	}
	e[i+1], e[hi] = e[hi], e[i+1]
	return i + 1
}

// quickselectTopK rearranges e in-place so that e[0..k-1] are the k highest-scoring
// entries. Average-case O(N), worst-case O(N²) but median-of-three avoids pathological cases.
func quickselectTopK(e []scoreEntry, k int) {
	if k <= 0 || len(e) == 0 {
		return
	}
	if k >= len(e) {
		return
	}
	lo, hi := 0, len(e)-1
	for lo < hi {
		p := partitionDesc(e, lo, hi)
		if p == k {
			return
		} else if p < k {
			lo = p + 1
		} else {
			hi = p - 1
		}
	}
}

// ─── ComputeEngine ─────────────────────────────────────────────────────────

// ComputeEngine orchestrates the two-phase compute cycle for one pod.
//
// Phase 1: C-kernel SIMD score update over the full PPSMatrix (int16, per-PPS).
// Phase 2: Goroutine-pool Quickselect ranking per pair using the PPS row score
// and BinTag eligibility from the inverted index.
type ComputeEngine struct {
	Matrix        *PPSMatrix
	Registry      *OrderRegistry
	Indexes       *InvertedIndexes
	Pairs         []PairInfo
	DB            *DoubleBuffer
	Topology      *TopologyState
	ItemDict      *MockItemDictClient
	numWorkers    int // goroutine pool size for Phase 2; set from ResourceConfig.NumWorkersVal
	scratch       []workerScratch
	pendingWarmup []PairInfo
}

// NewComputeEngine creates a ComputeEngine and allocates the PPSMatrix on the C heap.
// numWorkers sets the Phase 2 goroutine pool size; pass cfg.NumWorkersVal from
// initFromEnv to wire the resource config through explicitly. A value ≤ 0 falls
// back to the NumWorkers package default.
func NewComputeEngine(pairs []PairInfo, db *DoubleBuffer, topo *TopologyState, numWorkers int) *ComputeEngine {
	nw := numWorkers
	if nw <= 0 {
		nw = NumWorkers
	}
	scratch := make([]workerScratch, nw)
	for w := range scratch {
		scratch[w] = workerScratch{
			heuristic: make([]scoreEntry, 0, NumOrders),
			large:     make([]scoreEntry, 0, NumOrders),
			pbt:       make([]scoreEntry, 0, NumOrders),
			critical:  make([]scoreEntry, 0, NumOrders),
		}
	}
	return &ComputeEngine{
		Matrix:     NewPPSMatrix(numPPS),
		Registry:   NewOrderRegistry(),
		Indexes:    NewInvertedIndexes(),
		Pairs:      pairs,
		DB:         db,
		Topology:   topo,
		numWorkers: numWorkers,
		scratch:    scratch,
	}
}

// RunCycle executes one compute cycle with no ItemDict update (scores unchanged).
func (e *ComputeEngine) RunCycle() CycleStats {
	return e.RunCycleWithDict(nil)
}

// RunCycleWithDict executes one compute cycle.
//
//  1. Evict TTL-expired orders.
//  2. Phase 1: if update != nil, build a dense int32 delta slice and call CUpdateAllScores.
//  3. Phase 2: goroutine-pool Quickselect ranking per pair.
//  4. Publish the new ReadBuffer atomically.
//  5. Promote any pending warming-up pairs to Active.
func (e *ComputeEngine) RunCycleWithDict(update *ItemDictUpdate) CycleStats {
	start := time.Now()
	nowSecs := start.Unix()

	// Evict TTL-expired orders before computing new rankings.
	e.Registry.PurgeExpired(nowSecs, e.Indexes, e.Matrix)

	p1Ms := e.phase1Update(update)
	buf, p2Ms := e.phase2Rank(nowSecs)

	e.DB.Publish(buf)
	e.promoteWarmingPairs()

	totalMs := time.Since(start).Milliseconds()
	return CycleStats{
		Phase1Ms: p1Ms,
		Phase2Ms: p2Ms,
		TotalMs:  totalMs,
		SLAOk:    totalMs < 5_000,
	}
}

// HandleAddPair registers a new (PPS, BinTag) pair as WarmingUp and records it
// in pendingWarmup so it will be promoted after the next cycle.
func (e *ComputeEngine) HandleAddPair(ppsID, bintagID uint32) {
	e.Topology.AddWarmingUp(ppsID, bintagID)
	info := PairInfo{PPSID: ppsID, BinTagID: bintagID, Key: MakePairKey(ppsID, bintagID)}
	e.pendingWarmup = append(e.pendingWarmup, info)
	for _, p := range e.Pairs {
		if p.Key == info.Key {
			return
		}
	}
	e.Pairs = append(e.Pairs, info)
}

// HandleRemovePair marks a (PPS, BinTag) pair as Removed and removes it from
// the active pairs slice so it is excluded from future ranking cycles.
func (e *ComputeEngine) HandleRemovePair(ppsID, bintagID uint32) {
	key := MakePairKey(ppsID, bintagID)
	e.Topology.Remove(ppsID, bintagID)
	filtered := e.Pairs[:0]
	for _, p := range e.Pairs {
		if p.Key != key {
			filtered = append(filtered, p)
		}
	}
	e.Pairs = filtered
}

// HandleOrderRemoved evicts an order on a Kafka order-removed event.
func (e *ComputeEngine) HandleOrderRemoved(orderID uint32) {
	e.Registry.EvictOrder(orderID, e.Indexes, e.Matrix)
}

// phase1Update builds a dense int32 delta slice from the ItemDictUpdate then fans
// the PPSMatrix update across e.numWorkers goroutines, each calling the C SIMD
// kernel for a disjoint row slice. Returns the time spent in milliseconds.
func (e *ComputeEngine) phase1Update(update *ItemDictUpdate) int64 {
	t0 := time.Now()
	if update == nil || len(update.Updates) == 0 {
		return time.Since(t0).Milliseconds()
	}

	// Build int32 delta vector to avoid overflow from multiple i16-scaled contributions.
	delta := make([]int32, NumOrders)
	for _, u := range update.Updates {
		diff := u.NewContrib - u.OldContrib
		if diff == 0 {
			continue
		}
		diffI32 := int32(diff * ScoreScale)
		orderIDs, ok := e.Indexes.TPIDToOrders[u.TPID]
		if !ok {
			continue
		}
		for _, oid := range orderIDs {
			if oid < NumOrders {
				delta[oid] += diffI32
			}
		}
	}

	// Parallel Phase 1: distribute PPS rows across e.numWorkers goroutines.
	// Each goroutine calls the C SIMD kernel (CUpdateRowRange) for its disjoint
	// row slice — no shared writes between workers.
	numW := e.numWorkers
	if numW <= 0 {
		numW = NumWorkers
	}
	if numW > e.Matrix.NumPPS {
		numW = e.Matrix.NumPPS
	}
	rowsPerWorker := (e.Matrix.NumPPS + numW - 1) / numW

	var wg sync.WaitGroup
	for start := 0; start < e.Matrix.NumPPS; start += rowsPerWorker {
		end := min(start+rowsPerWorker, e.Matrix.NumPPS)
		wg.Add(1)
		go func(lo, hi int) {
			defer wg.Done()
			CUpdateRowRange(e.Matrix, lo, hi-lo, delta)
		}(start, end)
	}
	wg.Wait()

	return time.Since(t0).Milliseconds()
}

// ensureScratch grows the scratch bank to at least numW entries.
func (e *ComputeEngine) ensureScratch(numW int) {
	for len(e.scratch) < numW {
		e.scratch = append(e.scratch, workerScratch{
			heuristic: make([]scoreEntry, 0, NumOrders),
			large:     make([]scoreEntry, 0, NumOrders),
			pbt:       make([]scoreEntry, 0, NumOrders),
			critical:  make([]scoreEntry, 0, NumOrders),
		})
	}
}

// quickselectAndCopy partitions scratch so the top-TopK highest-scoring entries
// are at the front, sorts them descending, then writes their orderIDs to dst.
func quickselectAndCopy(dst []int32, scratch []scoreEntry) {
	k := min(len(scratch), TopK)
	if k == 0 {
		return
	}
	quickselectTopK(scratch, k)
	sort.Slice(scratch[:k], func(i, j int) bool {
		return scratch[i].score > scratch[j].score
	})
	for i := 0; i < k; i++ {
		dst[i] = scratch[i].orderID
	}
}

// phase2Rank runs the Phase 2 parallel ranking and returns a freshly populated
// ReadBuffer plus the time spent in milliseconds.
//
// Algorithm: static work partitioning (no channel) — each goroutine owns a
// contiguous slice of pairs and a pre-allocated workerScratch bank entry.
// A single pass over eligible orders per pair fills all four scratch slices
// simultaneously, reducing OrderMeta cache misses 4× vs the old 4-pass approach.
func (e *ComputeEngine) phase2Rank(nowSecs int64) (*ReadBuffer, int64) {
	t0 := time.Now()
	numPairs := len(e.Pairs)
	buf := NewReadBuffer(numPairs)

	if numPairs == 0 {
		return buf, time.Since(t0).Milliseconds()
	}

	critDeadline := nowSecs + CriticalCutoff

	numW := e.numWorkers
	if numW <= 0 {
		numW = NumWorkers
	}
	if numW > numPairs {
		numW = numPairs
	}
	e.ensureScratch(numW)

	pairsPerWorker := (numPairs + numW - 1) / numW

	var wg sync.WaitGroup
	for w := 0; w < numW; w++ {
		lo := w * pairsPerWorker
		hi := min(lo+pairsPerWorker, numPairs)
		if lo >= numPairs {
			break
		}
		wg.Add(1)
		go func(workerID, start, end int) {
			defer wg.Done()
			sc := &e.scratch[workerID]

			for p := start; p < end; p++ {
				pair := e.Pairs[p]
				// PPS IDs are 0-based in go-cgo.
				ppsRow := e.Matrix.Row(int(pair.PPSID))
				eligible := e.Indexes.BinTagToOrders[pair.BinTagID]

				// ── Single pass: fill all four scratch slices simultaneously ──
				sc.heuristic = sc.heuristic[:0]
				sc.large = sc.large[:0]
				sc.pbt = sc.pbt[:0]
				sc.critical = sc.critical[:0]

				for _, oid := range eligible {
					if oid >= NumOrders {
						continue
					}
					meta := &e.Registry.Meta[oid]
					if !meta.Active {
						continue
					}
					score := int32(ppsRow[oid])
					sc.heuristic = append(sc.heuristic, scoreEntry{score, int32(oid)})
					if meta.RequiredQty >= LargeQtyThreshold {
						sc.large = append(sc.large, scoreEntry{score, int32(oid)})
					}
					if meta.PBTDeadline > 0 {
						dl := scoreEntry{int32(nowSecs - meta.PBTDeadline), int32(oid)}
						sc.pbt = append(sc.pbt, dl)
						if meta.PBTDeadline <= critDeadline {
							sc.critical = append(sc.critical, dl)
						}
					}
				}

				quickselectAndCopy(buf.ListMut(p, TblHeuristic), sc.heuristic)
				quickselectAndCopy(buf.ListMut(p, TblLarge), sc.large)
				quickselectAndCopy(buf.ListMut(p, TblPBT), sc.pbt)
				quickselectAndCopy(buf.ListMut(p, TblCritical), sc.critical)
			}
		}(w, lo, hi)
	}
	wg.Wait()

	return buf, time.Since(t0).Milliseconds()
}

// promoteWarmingPairs promotes all pairs that were warming up when the last cycle
// started to StatusActive, and clears the pendingWarmup list.
func (e *ComputeEngine) promoteWarmingPairs() {
	for _, p := range e.pendingWarmup {
		e.Topology.Activate(p.PPSID, p.BinTagID)
	}
	e.pendingWarmup = e.pendingWarmup[:0]
}

File: rust-impl/src/compute.rs
use crate::mock::{ItemDictUpdate, MockItemDictClient};
use crate::registry::{InvertedIndexes, OrderRegistry};
use crate::topology::TopologyState;
use crate::types::*;
use rayon::prelude::*;
use std::cell::RefCell;
use std::sync::Arc;
use std::time::Instant;

// ─── Thread-local scratch buffer ─────────────────────────────────────────────

struct WorkerScratch {
    heuristic: Vec<SEntry>,
    large:     Vec<SEntry>,
    pbt:       Vec<SEntry>,
    critical:  Vec<SEntry>,
}

thread_local! {
    static SCRATCH: RefCell<WorkerScratch> = RefCell::new(WorkerScratch {
        heuristic: Vec::with_capacity(NUM_ORDERS),
        large:     Vec::with_capacity(NUM_ORDERS),
        pbt:       Vec::with_capacity(NUM_ORDERS),
        critical:  Vec::with_capacity(NUM_ORDERS),
    });
}

// ─── ComputeEngine ───────────────────────────────────────────────────────────

pub struct ComputeEngine {
    /// [NUM_PPS × NUM_ORDERS] i16 score matrix; row = pps_id.
    pub matrix: PPSMatrix,
    pub registry: OrderRegistry,
    pub indexes: InvertedIndexes,
    pub pairs: Vec<PairInfo>,
    pub db: Arc<DoubleBuffer>,
    pub topology: Arc<TopologyState>,
    pub item_dict_client: MockItemDictClient,
    pending_warmup: Vec<(u32, u32)>,
}

impl ComputeEngine {
    pub fn new(
        pairs: Vec<PairInfo>,
        db: Arc<DoubleBuffer>,
        topology: Arc<TopologyState>,
    ) -> Self {
        Self {
            matrix: PPSMatrix::new(NUM_PPS),
            registry: OrderRegistry::new(),
            indexes: InvertedIndexes::new(),
            pairs,
            db,
            topology,
            item_dict_client: MockItemDictClient::new(42, 100_000),
            pending_warmup: Vec::new(),
        }
    }

    pub fn run_cycle(&mut self) -> CycleStats {
        let item_dict = self.item_dict_client.fetch();
        self.run_cycle_with_dict(&item_dict)
    }

    pub fn run_cycle_with_dict(&mut self, item_dict: &ItemDictUpdate) -> CycleStats {
        let cycle_start = Instant::now();
        let now_secs = current_unix_secs();

        // Evict orders whose TTL has expired before computing new rankings.
        self.registry
            .purge_expired(now_secs, &mut self.indexes, &mut self.matrix);

        let d1 = self.phase1_update(item_dict);
        let (new_buf, d2) = self.phase2_rank(now_secs);

        self.db.publish(new_buf);
        self.promote_warming_pairs();

        let total_ms = cycle_start.elapsed().as_millis() as u64;
        CycleStats {
            phase1_ms: d1.as_millis() as u64,
            phase2_ms: d2.as_millis() as u64,
            total_ms,
            sla_ok: total_ms < 5_000,
        }
    }

    // ── Phase 1 ──────────────────────────────────────────────────────────────

    fn phase1_update(&mut self, item_dict: &ItemDictUpdate) -> std::time::Duration {
        let t0 = Instant::now();

        // Accumulate a single i32 delta vector across all TPID updates.
        // i32 avoids overflow when summing multiple i16-scaled deltas.
        let mut delta = vec![0i32; NUM_ORDERS];

        for update in &item_dict.updates {
            let diff = update.new_contrib - update.old_contrib;
            if diff == 0.0 {
                continue;
            }
            let diff_i16 = (diff * SCORE_SCALE) as i32;
            if let Some(order_ids) = self.indexes.tpid_to_orders.get(&update.tpid) {
                for &oid in order_ids {
                    if (oid as usize) < NUM_ORDERS {
                        delta[oid as usize] += diff_i16;
                    }
                }
            }
        }

        // Apply delta to every PPS row in parallel.
        // Only NUM_PPS rows (e.g. 50) × NUM_ORDERS — much smaller than the old
        // num_pairs × NUM_ORDERS layout.
        self.matrix
            .data
            .par_chunks_exact_mut(NUM_ORDERS)
            .for_each(|row| {
                for (s, &d) in row.iter_mut().zip(delta.iter()) {
                    *s = s.saturating_add(d.clamp(i16::MIN as i32, i16::MAX as i32) as i16);
                }
            });

        t0.elapsed()
    }

    // ── Phase 2 ──────────────────────────────────────────────────────────────

    fn phase2_rank(&self, now_secs: i64) -> (ReadBuffer, std::time::Duration) {
        let t0 = Instant::now();
        let num_pairs = self.pairs.len();
        let mut write_buf = ReadBuffer::new(num_pairs);

        let critical_deadline = now_secs + CRITICAL_CUTOFF_SECS;
        let matrix_ref: &PPSMatrix = &self.matrix;
        let registry_ref: &OrderRegistry = &self.registry;
        let indexes_ref: &InvertedIndexes = &self.indexes;
        let pairs_ref: &[PairInfo] = &self.pairs;

        write_buf
            .data
            .par_chunks_exact_mut(NUM_TABLES * TOP_K)
            .enumerate()
            .for_each(|(pair_idx, chunk)| {
                if pair_idx >= pairs_ref.len() {
                    return;
                }
                let pair = &pairs_ref[pair_idx];
                let pps_row = matrix_ref.row(pair.pps_id as usize);
                let eligible: &[u32] = indexes_ref
                    .bintag_to_orders
                    .get(&pair.bintag_id)
                    .map(|v| v.as_slice())
                    .unwrap_or(&[]);

                SCRATCH.with(|cell| {
                    let mut sc = cell.borrow_mut();

                    // ── Single pass: fill all four scratch slices simultaneously ──
                    sc.heuristic.clear();
                    sc.large.clear();
                    sc.pbt.clear();
                    sc.critical.clear();

                    for &oid in eligible {
                        let meta = &registry_ref.meta[oid as usize];
                        if !meta.active {
                            continue;
                        }
                        let score = pps_row[oid as usize] as i32;
                        sc.heuristic.push(SEntry { score, id: oid as i32 });
                        if meta.required_qty > LARGE_QTY_THRESHOLD {
                            sc.large.push(SEntry { score, id: oid as i32 });
                        }
                        if meta.pbt_deadline > 0 {
                            let dl_score = (now_secs - meta.pbt_deadline) as i32;
                            sc.pbt.push(SEntry { score: dl_score, id: oid as i32 });
                            if meta.pbt_deadline <= critical_deadline {
                                sc.critical.push(SEntry { score: dl_score, id: oid as i32 });
                            }
                        }
                    }

                    fill_topk_desc(chunk, TBL_HEURISTIC, &mut sc.heuristic);
                    fill_topk_desc(chunk, TBL_LARGE, &mut sc.large);
                    fill_topk_desc(chunk, TBL_PBT, &mut sc.pbt);
                    fill_topk_desc(chunk, TBL_CRITICAL, &mut sc.critical);
                });
            });

        (write_buf, t0.elapsed())
    }

    // ── Warmup ────────────────────────────────────────────────────────────────

    fn promote_warming_pairs(&mut self) {
        for (pps_id, bintag_id) in self.pending_warmup.drain(..) {
            self.topology.activate(pps_id, bintag_id);
        }
    }

    // ── Topology handlers ─────────────────────────────────────────────────────

    pub fn handle_add_pair(&mut self, pps_id: u32, bintag_id: u32) {
        self.topology.add_warming_up(pps_id, bintag_id);
        self.pending_warmup.push((pps_id, bintag_id));

        let key = make_pair_key(pps_id, bintag_id);
        let already_tracked = self.pairs.iter().any(|p| p.key == key);
        if !already_tracked {
            self.pairs.push(PairInfo { pps_id, bintag_id, key });
            // Grow PPSMatrix if this pps_id exceeds current row count.
            let required = pps_id as usize + 1;
            if required > self.matrix.num_pps {
                let extra_rows = required - self.matrix.num_pps;
                self.matrix.data.extend(vec![0i16; extra_rows * NUM_ORDERS]);
                self.matrix.num_pps = required;
            }
        }
    }

    pub fn handle_remove_pair(&mut self, pps_id: u32, bintag_id: u32) {
        self.topology.remove(pps_id, bintag_id);
        self.pending_warmup.retain(|&(p, b)| !(p == pps_id && b == bintag_id));
        // Remove from pairs list.
        let key = make_pair_key(pps_id, bintag_id);
        self.pairs.retain(|p| p.key != key);
    }

    /// Evict an order on a Kafka removal event.
    pub fn handle_order_removed(&mut self, order_id: u32) {
        self.registry
            .evict_order(order_id, &mut self.indexes, &mut self.matrix);
    }
}

// ─── Quickselect helper ──────────────────────────────────────────────────────

/// Partially sort `scratch` so the top-k highest-scoring entries are in
/// scratch[0..k], then sort that prefix descending, and write to chunk[table*TOP_K..].
fn fill_topk_desc(chunk: &mut [i32], table: usize, scratch: &mut Vec<SEntry>) {
    let k = scratch.len().min(TOP_K);
    if k == 0 {
        return;
    }
    if scratch.len() > k {
        scratch.select_nth_unstable_by(k - 1, |a, b| b.score.cmp(&a.score));
    }
    scratch[..k].sort_unstable_by(|a, b| b.score.cmp(&a.score));
    let slot = &mut chunk[table * TOP_K..(table + 1) * TOP_K];
    for (i, e) in scratch[..k].iter().enumerate() {
        slot[i] = e.id;
    }
}

// ─── Helper ──────────────────────────────────────────────────────────────────

pub fn current_unix_secs() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}
